File: problems-adapt/0100-0200/0105_nth-five-smooth-number/solution_heap.py
import heapq


class Solution:
    def nthFiveSmooth(self, n: int) -> int:
        # Frontier of the generation process: a min-heap seeded with 1, so
        # the smallest not-yet-emitted five-smooth number is always at its top.
        heap = [1]
        # The heap is a frontier, not a set: pushing every successor would
        # enqueue duplicates (6 = 2·3 = 3·2), so seen gates each push.
        seen = {1}
        for _ in range(n - 1):
            value = heapq.heappop(heap)
            for factor in (2, 3, 5):
                multiple = value * factor
                if multiple not in seen:
                    seen.add(multiple)
                    heapq.heappush(heap, multiple)
        # After n-1 pops the heap top is the n-th five-smooth number in order.
        return heap[0]

File: problems/0201-0300/0264_nth-five-smooth-number/solution_heap.go
import (
	"container/heap"
)

// A min-heap of int64: the frontier of the five-smooth generation process.
type smoothHeap []int64

func (h smoothHeap) Len() int           { return len(h) }
func (h smoothHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h smoothHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *smoothHeap) Push(x any) { *h = append(*h, x.(int64)) }

func (h *smoothHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func nthFiveSmooth(n int) int {
	// Seeded with 1, the heap top is always the smallest not-yet-emitted
	// five-smooth number. 64-bit elements: pushed multiples can overshoot int32.
	h := &smoothHeap{1}
	// The heap is a frontier, not a set: pushing every successor would
	// enqueue duplicates (6 = 2·3 = 3·2), so seen gates each push.
	seen := map[int64]bool{1: true}
	for i := 1; i < n; i++ {
		value := heap.Pop(h).(int64)
		for _, factor := range []int64{2, 3, 5} {
			multiple := value * factor
			if !seen[multiple] {
				seen[multiple] = true
				heap.Push(h, multiple)
			}
		}
	}
	// After n-1 pops the heap top is the n-th five-smooth number in order.
	return int((*h)[0])
}
